src/Raven.Server/Documents/CountersStorage.cs

```diff
@@ -1820,6 +1911,243 @@ public string DeleteCounter(DocumentsOperationContext context, string documentId
+            }
+        }
+
+        public long PurgeCountersAndCounterTombstones(DocumentsOperationContext context, string collection, long upto, long numberOfEntriesToDelete)
+        {
+            var collectionName = _documentsStorage.GetCollection(collection, throwIfDoesNotExist: false);
+            if (collectionName == null)
+                return 0;
+
+            var deletedCounters = PurgeCounters(upto, context, collectionName, numberOfEntriesToDelete);
+            var deletedTombstoneEntries = PurgeCounterTombstones(upto, context, collectionName, numberOfEntriesToDelete - deletedCounters);
+            return deletedTombstoneEntries + deletedCounters;
+        }
+
+        private long PurgeCounters(long upto, DocumentsOperationContext context, CollectionName collectionName, long numberOfEntriesToDelete)
+        {
+            var counterTombstonesTableName = collectionName.GetTableName(CollectionTableType.CounterTombstones);
+            var tombstonesTable = context.Transaction.InnerTransaction.OpenTable(CounterTombstonesSchema, counterTombstonesTableName);
+            if (tombstonesTable == null || tombstonesTable.NumberOfEntries == 0 || numberOfEntriesToDelete <= 0)
+                return 0;
+
+            var table = GetCountersTable(context.Transaction.InnerTransaction, collectionName);
+            if (table == null || table.NumberOfEntries == 0)
+                return 0;
+
+            var deleted = 0L;
+
+            foreach (var item in GetCounterTombstonesFrom(context, 0, upto))
+            {
+                // first, we extract the document Id and the deleted counter name from the counter tombstones entry
+                // then, we extract the data from the counters table, by the counter group key
+                // and remove the deleted counter (change the data)
+                // at the end, we'll write the new data to disk and update the document
+
+                if (item.Etag > upto || numberOfEntriesToDelete <= deleted)
+                    return deleted;
+
+                using (DocumentIdWorker.GetSliceFromId(context, item.DocumentId, out Slice documentKeyPrefix, separator: SpecialChars.RecordSeparator))
+                using (DocumentIdWorker.GetLower(context.Allocator, context.GetLazyString(item.Name), out Slice counterNameSlice))
+                using (context.Allocator.Allocate(documentKeyPrefix.Size + counterNameSlice.Size, out var counterKeyBuffer))
+                using (CreateCounterKeySlice(context, counterKeyBuffer, documentKeyPrefix, counterNameSlice, out var counterKeySlice))
+                {
+                    if (table.SeekOneBackwardByPrimaryKeyPrefix(documentKeyPrefix, counterKeySlice, out var existing) == false)
+                        return deleted;
+
+                    BlittableJsonReaderObject data;
+                    using (data = GetCounterValuesData(context, ref existing))
+                    {
+                        data = data.Clone(context);
```

Comment:
why do we need to clone this so early? do we need to clone it at all?

Reply:
Yes. It seems we must clone the data there (as in `PutOrIncrement` method)